src/main/java/dev/morling/onebrc/CalculateAverage_royvanrijn.java

```diff
@@ -58,210 +57,173 @@
-
-                // Calculate using native ordering (fastest?):
-                bb.order(ByteOrder.nativeOrder());
-
-                // Record the order it is and calculate accordingly:
-                final boolean bufferIsBigEndian = bb.order().equals(ByteOrder.BIG_ENDIAN);
-                MeasurementRepository measurements = new MeasurementRepository();
-
-                int startPointer;
-                int limit = bb.limit();
-                while ((startPointer = bb.position()) < limit) {
-
-                    int delimiterPointer, endPointer;
-
-                    // SWAR method to find delimiter *and* record the cityname as long[] *and* calculate a hash:
-                    findNextDelimiterAndCalculateHash(bb, SEPARATOR_PATTERN, startPointer, limit, delimiterPointerAndHash, cityNameAsLongArray, bufferIsBigEndian);
-                    delimiterPointer = delimiterPointerAndHash[0];
-
-                    // Simple lookup is faster for '\n' (just three options)
-                    if (delimiterPointer >= limit) {
-                        return measurements;
-                    }
-                    // Extract the measurement value (10x):
-                    final int cityNameLength = delimiterPointer - startPointer;
-
-                    int measuredValue;
-                    int neg = 1;
-                    if (bb.get(++delimiterPointer) == '-') {
-                        neg = -1;
-                        delimiterPointer++;
-                    }
-                    byte dot;
-                    if ((dot = (bb.get(delimiterPointer + 1))) == '.') {
-                        measuredValue = neg * ((bb.get(delimiterPointer)) * 10 + (bb.get(delimiterPointer + 2)) - 528);
-                        endPointer = delimiterPointer + 3;
-                    }
-                    else {
-                        measuredValue = neg * (bb.get(delimiterPointer) * 100 + dot * 10 + bb.get(delimiterPointer + 3) - 5328);
-                        endPointer = delimiterPointer + 4;
-                    }
-
-                    // Store everything in a custom hashtable:
-                    measurements.update(cityNameAsLongArray, bb, cityNameLength, delimiterPointerAndHash[1]).updateWith(measuredValue);
-
-                    bb.position(endPointer + 1); // skip to next line.
+    public void run() throws Exception {
+
+        // Calculate input segments.
+        int numberOfChunks = Runtime.getRuntime().availableProcessors();
+        long[] chunks = getSegments(numberOfChunks);
+
+        // Parallel processing of segments.
+        TreeMap<String, Measurement> results = IntStream.range(0, chunks.length - 1)
+                .mapToObj(chunkIndex -> process(chunks[chunkIndex], chunks[chunkIndex + 1])).parallel()
+                .flatMap(MeasurementRepository::get)
+                .collect(Collectors.toMap(e -> e.city, MeasurementRepository.Entry::measurement, Measurement::updateWith, TreeMap::new));
```

Comment:
Lambdas are not super good for startup, but maybe with CDS isn't like that! Anonymous classes tends to have more  predictable costs (lower usually)

Reply:
Yes. And that loading is single-threaded. This is why it should be a bigger benefit in terms of percentage of execution time on a stronger machine to create a native image. Overall, the loaded and compiled code is still relatively minimal compared to what is loaded in your typical Spring / Quarkus / Micronaut application ;-).